peer/bind.go

```diff
@@ -0,0 +1,120 @@
+// Copyright (c) 2017 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package peer
+
+import (
+	"context"
+
+	"go.uber.org/yarpc/api/peer"
+	"go.uber.org/yarpc/api/transport"
+	"go.uber.org/yarpc/internal/errors"
+	"go.uber.org/yarpc/internal/introspection"
+	intsync "go.uber.org/yarpc/internal/sync"
+)
+
+// Binder is a function, provided to peer.Bind, that accepts a peer list and binds
+// it to a peer provider for the duration of the returned lifecycle.
+type Binder func(peer.List) transport.Lifecycle
+
+// Bind couples a peer list with a peer provider.
+// The peer provider must produce a binder that takes the peer list and returns
+// a lifecycle for the duration of the binding.
+func Bind(chooserList peer.ChooserList, bind Binder) *BoundChooser {
+	return &BoundChooser{
+		once:    intsync.Once(),
+		binding: bind(chooserList),
+		chooser: chooserList,
+	}
+}
+
+// BoundChooser is a peer chooser that couples a peer list and a peer provider
+// for the duration of its lifecycle.
+type BoundChooser struct {
+	once    intsync.LifecycleOnce
+	binding transport.Lifecycle
+	chooser peer.Chooser
+}
+
+// Choose returns a peer from the bound peer list.
+func (c *BoundChooser) Choose(ctx context.Context, treq *transport.Request) (peer peer.Peer, onFinish func(error), err error) {
+	return c.chooser.Choose(ctx, treq)
+}
+
+// Start starts the peer list and the peer provider binding.
+func (c *BoundChooser) Start() error {
+	return c.once.Start(c.start)
+}
+
+func (c *BoundChooser) start() error {
+	var errors errors.ErrorGroup
+
+	if err := c.chooser.Start(); err != nil {
+		errors = append(errors, err)
+	}
+
+	if err := c.binding.Start(); err != nil {
```

Comment:
if the chooser fails but the binding starts won't that have weird transient issues?  Should we short circuit if the chooser doesn't start properly?